### DISCUSSION TITLE
Some kind of Fixed Joint behavior or reduction of 'springyness' from multiple 6 DOF Joints chained together.

### DISCUSSION THREAD

**Discussion by caspercasanova** [godot-jolt/godot-jolt#1010] [Questions] [answered]
Hey love the project and the work that has/is being done. 

This might be the wrong channel, but I was wondering about the potential for a Fixed Joint in Jolt? I have been attempting to use the `6 DOF Jolt Joint` as a way to Pin different `Rigid Bodies` to a `Main Rigid Body` (in my case a Helicopter and different sections of the tail boom). 

I have been tinkering around and found that as these sections get longer, with more joints chained together (beyond 2/3) things start acting a bit _"springy"_.

I changed the weight of the rigid bodies which had a dampening effect which made the _'springiness'_ less noticeable, but I am wondering if there is a proper way to have a "Fixed" Joint. 

I am sure this is probably answered with [this discussion post](https://github.com/godot-jolt/godot-jolt/discussions/999#discussion-7312808), but I am just wondering if there is a way to maybe 'hack' it or even better maybe I am just flat out wrong on how I'm using the `6DOF Joint`? Hoping to hear your opinions on this.

Really trying to avoid needing to use a ton of Area3D's; it would make Blender -> Godot easier I believe. 

Here is the [repo](https://github.com/caspercasanova/helicopter) I have been working on, a little helicopter game. "WASD+ Shift & Ctrl for up and down & F for a grappling hook". 

Thanks. 

And yes I am using the latest Jolt from a couple days ago. 

Some notes I found mentioning "Fixed Joints" (I know they aren't really applicable here, but figured I'd share)
- https://ezengine.net/pages/docs/physics/jolt/constraints/jolt-fixed-constraint-component.html
- https://docs.nvidia.com/gameworks/content/gameworkslibrary/physx/guide/Manual/Joints.html

**Comment by mihe** [comment, marked as answer]
I do already make use of Jolt's `FixedConstraint` for when you have a `HingeJoint3D` or `SliderJoint3D` and set both the upper and lower limits of the joint to zero, and leave any spring properties disabled. I can't remember right now why I didn't do this for the other joints as well, but I vaguely recall there being some reason for that.

Note that even when using a `HingeJoint3D` or `SliderJoint3D` with a limit of zero (and unsprung) the joint will only be as "fixed" as the physics timestep and the number of solver iterations allow for. While you can tune both of these things _globally_ (which I wouldn't recommend) there is unfortunately no way to tune the number of solver iterations per-joint with the default Godot joints, but I have exposed this on the Jolt-specific substitute joints, meaning `JoltHingeJoint3D` and `JoltSliderJoint3D`.

So if you use either a `JoltHingeJoint3D` or `JoltSliderJoint3D`, while similarly setting the joint limits to zero (and leaving it unsprung), you can bump up the "Velocity Iterations" and "Position Iterations" properties (under the "Solver Overrides" group) for each joint to give them much more rigidity, at the cost of performance.

EDIT: It's probably worth mentioning as well that (from what I understand) these solver iterations are effectively "viral", so whatever non-static body that your joint is connected to, and any non-static body that they in turn come in contact with, and so forth, will have these number of solver iterations applied to them. So it can get quite expensive if you have a convoluted arrangement of joints and dynamic bodies for example.

**Reply by caspercasanova** [reply]
Thanks for the response. Super interesting. I'll experiment with those (it's funny those were the ones I used first, but didn't think to increase the Velocity + Position Iterations). (Also I probably should have just "Ctrl+F+'Fixed Constraint'" in the Godot Jolt repo to find my answer. 

I think I can do some tricks to make the vehicle feel destructible while keeping the joint chaining to a minimum (I suppose it's ultimately my goal). Ultimately I probably just need to really dig into the Godot Jolt + Jolt repos. It seems there are a lot of Foot Guns that could go unnoticed for a novice game dev. 

Thanks once again.

**Reply by eSavag** [reply]
I'm having a lot of trouble with this having moved from the Godot jolt extension to the inbuilt version of jolt in 4.5. I have two sets of raycast-based track systems on a tank, and they are now constantly spinning out or lagging behind the main vehicle body when forces are applied. I didn't have this problem with the extension - did something change, or is there a new best practice for setting up fixed joint behaviour? this is affecting hinge, slider and dof joints, including those setup as described above.

**Reply by mihe** [reply]
@eSavag If there's a significant discrepancy between the extension and the built-in module in 4.5, without any changes to the scene itself, then I would recommend that you [create an issue](https://github.com/godotengine/godot/issues/new?template=bug_report.yml) for it in the Godot repository.

As far as I know, pretty much nothing about the joints has changed from the extension to the built-in module on the integration side, other than the fact that we don't have things like the above mentioned solver overrides, but the built-in module does also run a newer version of Jolt itself, so it could be a regression there I suppose.

**Comment by Misjre** [comment]
I'm also having tons of issues and inconsistencies with this, I was able to replicate a 4 bar linkage in one project without issues at all, and now I can't create a similar suspension system.... It has a chain of 3 bodies in total, all pin joints except the rear wheel, which is the last on the chain. It is not rigid at all, extremely soft, even when I have the solver at high values and the physics ticks at 240hz

**Reply by caspercasanova** [reply]
I've returned to my example as well. Tried all kinds of joints and different settings and still get a softness as well. I haven't dug into it for a while, but would be interested in looking to how to fix this. Perhaps the best way might be to just make a proper pr for a fixed joint of some kind.

**Reply by mihe** [reply]
> Perhaps the best way might be to just make a proper pr for a fixed joint of some kind.

If you're talking about a PR to this extension, note that it's in maintenance mode now and I will not merge any new features or other "enhancements" into it.

If you're talking about a PR to Godot itself, there already is one: godotengine/godot#101575

That PR is unlikely to see any movement until at least Jolt is made the default 3D physics engine in Godot though.

But again, as mentioned above, I already make use of Jolt's `FixedConstraint` under the hood for `HingeJoint3D` and `SliderJoint3D` (and the `Jolt*` equivalents found in this extension) if you set both limits to the same value and leave the spring properties at their default values. If you're not seeing any benefit from that then exposing it as its own joint won't make any difference.

@Leteen It's unclear what "extremely soft" means exactly, so I would recommend you create a minimal reproduction project and create a new issue in this repository. That way we can see if it's an actual bug or just a known shortcoming with the type of solver Jolt uses.
